tidy3d/components/medium.py

```diff
@@ -1079,206 +736,631 @@ class Drude(DispersiveMedium):
+    >>> eps_components = {f"eps_{d}{d}": eps_diagonal_data for d in "xyz"}
+    >>> dataset = PermittivityDataset(**eps_components)
+    >>> dielectric = CustomMedium(dataset=dataset, name='my_medium')
+    >>> eps = dielectric.eps_model(200e12)
     """
 
-    xx: IsotropicMediumType = pd.Field(
-        ...,
-        title="XX Component",
-        description="Medium describing the xx-component of the diagonal permittivity tensor.",
-        discriminator=TYPE_TAG_STR,
-    )
-
-    yy: IsotropicMediumType = pd.Field(
-        ...,
-        title="YY Component",
-        description="Medium describing the yy-component of the diagonal permittivity tensor.",
-        discriminator=TYPE_TAG_STR,
-    )
-
-    zz: IsotropicMediumType = pd.Field(
+    dataset: DispersiveDataset = pd.Field(
         ...,
-        title="ZZ Component",
-        description="Medium describing the zz-component of the diagonal permittivity tensor.",
-        discriminator=TYPE_TAG_STR,
+        title="Dispersion Coefficients Dataset",
+        description="User-supplied dataset containing the spatially-varying coefficients for "
+        "Defining the medium properties. ",
     )
 
-    @cached_property
-    def components(self) -> Dict[str, Medium]:
-        """Dictionary of diagonal medium components."""
-        return dict(xx=self.xx, yy=self.yy, zz=self.zz)
-
-    @cached_property
-    def n_cfl(self):
-        """This property computes the index of refraction related to CFL condition, so that
-        the FDTD with this medium is stable when the time step size that doesn't take
-        material factor into account is multiplied by ``n_cfl``.
-
-        For this medium, it takes the minimal of ``n_clf`` in all components.
-        """
-        return min((mat_component.n_cfl for mat_component in self.components.values()))
-
-    @ensure_freq_in_range
-    def eps_model(self, frequency: float) -> complex:
-        """Complex-valued permittivity as a function of frequency."""
-
-        return np.mean(self.eps_diagonal(frequency), axis=0)
-
-    @ensure_freq_in_range
-    def eps_diagonal(self, frequency: float) -> Tuple[complex, complex, complex]:
-        """Main diagonal of the complex-valued permittivity tensor as a function of frequency."""
-
-        eps_xx = self.xx.eps_model(frequency)
-        eps_yy = self.yy.eps_model(frequency)
-        eps_zz = self.zz.eps_model(frequency)
-        return (eps_xx, eps_yy, eps_zz)
+    # TODO: need to implement ABC methods to handle SpatialDataArray values?
+    # @cached_property
+    # def pole_residue(self):
+    #     """Representation of Medium as a pole-residue model."""
+    #     # TODO: generalize the pole_residue method of the DispersiveModel to
+    #     # # basic idea
+    #     # eps_inf = self.dataset.eps_inf
+    #     # coeffs = self.dataset.coeffs
+    #     # values = Drude.eps_model()
+    #     #     dataset = dataset.sel(points)
+    #     #     value = Drude.eps_model(dataset.values)
```

Comment:
A general python question here: can you call `value = Drude.eps_model(frequency)` here, or is a Drude object constructed?

Reply:
Yea I think this should be possible. But since it's a method, it expects `self` so I'm pretty sure you'd need to pass an instance that can act similar to a Drude as first argument like `value = Drude.eps_model(self, frequency)`